src/data/multiset/basic.lean

```diff
@@ -481,6 +481,13 @@ def card : multiset α →+ ℕ :=
 
 @[simp] theorem coe_card (l : list α) : card (l : multiset α) = length l := rfl
 
+@[simp] theorem length_to_list (s : multiset α) : s.to_list.length = s.card :=
+by rw [← coe_card, coe_to_list]
+
+/-- Index the elements of a multiset `s` using a function from the finite type `fin s.card`. -/
+noncomputable def of_fin_card (s : multiset α) (n : fin s.card) : α :=
```

Comment:
The full name will be `multiset.of_fin_card` right? It seems a little weird...

Reply:
Seems I can't, so how about renaming to `multiset.fin_card_index_fun`?

Reply:
I am maybe misunderstanding something, but why not having something like `multiset.nth_le_card` (similar to [nth_le](https://leanprover-community.github.io/mathlib_docs/init/data/list/basic.html#list.nth_le))? Also I don't understand what do you mean by "using a function from the finite type `fin s.card`". This is just the `n`-th element of `s` after having chosen a random order on the multiset, isn't it?

Reply:
I find it a bit weird to talk about the "nth" element of a multiset, since it depends on an arbitrary choice of an order. To me `of_fin_card` is just an auxiliary function to transfer results about fintype/finset-indexed big operators to/from results about big operators over multiset (using `finset.prod_eq_multiset_prod`), and any fintype works, as long as it satisfies the property `map_of_fin_card` (in the Zulip thread there are two or three constructions, but not sure if the property is verified).

I defined `of_fin_card` as a function from `fin s.card` instead of mimicking `list.nth_le`, because the first is what I really want, and there doesn't seem to be much API that can be developed in parallel to `list.nth_le`, because of the arbitrary choice. If we define `multiset.nth_le`, then `of_fin_card` would be stated in terms of `λ n, s.nth_le n n.2` instead of `s.of_fin_card`, which seems fine too. What do you think?

Reply:
I'm not really comfortable with this function -- the effect of this is that it allows users of it to avoid proving that things are independent of the choice of the multiset's representative. The "right" way is to use `quot.rec_on m` to get a list representative, which requires that you prove the result is independent of the list.

This was one of the reasons I [had proposed](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/.E2.9C.94.20fintype.20enumerating.20a.20multiset/near/286819701) this type as an indexing type:
```
Σ (x : m.to_finset), fin (m.count x)
```
It's also conducive to computations using `#eval`. (An alternative approach is coercing the `finset` in [this post](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/.E2.9C.94.20fintype.20enumerating.20a.20multiset/near/286826672) to `Type`, and I suspect it would give a better API.)

Reply:
Notice that `multiset.to_list` already uses choice, and people have been using [multiset.prod_to_list](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/multiset.html#multiset.prod_to_list) etc. to derive results about multiset.prod/sum from those about list.prod/sum, and what I do here is immediately similar to that (and even uses `multiset.to_list` to define `of_fin_card`).

Your approach is also not completely satisfactory because both `to_finset` and `count` require `decidable_eq`, and hence #eval requires it to work; however it seems hopeless to define an enumeration function computably without decidable_eq. If you have a computable multiset, there should be a computable underlying list, and you can use `list.nth_le` to enumerate it (without decidable_eq). Unfortunately this enumeration is not invariant under permutation, so we don't get a computable enumeration for a multiset.

If we care about computability, how about defining `of_fin_card` to take arguments `{l : list α} (hl : ⟦l⟧ = s)`? This way we can use `hl := rfl` when `s` comes from a (computable) list, and `hl := s.coe_to_list` for general `s`. `map_of_fin_card` would also take the two arguments and the proof could still use `map_nth_le` (a lemma about lists, which should be in mathlib IMO independent of what enumeration function we choose for multisets), and we may provide a version of `map_of_fin_card` with `hl := s.coe_to_list` for convenience. If we adopt your approach, the analogue of `map_of_fin_card` must be proved another way.

Alternatively, if you just don't want the definition `of_fin_card` in mathlib, I could simply inline the definition `of_fin_card` into `map_of_fin_card`, but that seems less favorable than the solution in the last paragraph.

Reply:
> and people have been using [multiset.prod_to_list](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/multiset.html#multiset.prod_to_list) etc. to derive results about multiset.prod/sum from those about list.prod/sum

A lot of the time people use these lemmas only because they are not aware of the appropriate induction principle. Could you point to an example of where you've seen this?

Reply:
> To me `of_fin_card` is just an auxiliary function to trans fer results about fintype/finset-indexed big operators to/from results about big operators over multiset (using `finset.prod_eq_multiset_prod`),

I worry that this approach makes things convoluted; normally we transfer results from `list`s to `multiset`s to `finset`s, not the other way around.
I can see us ending up with weird import constraints going down this path.

Reply:
[Searching for `prod_to_list`](https://github.com/leanprover-community/mathlib/search?q=prod_to_list) in the mathlib repo and just two use cases turns up: in [multiset.periodic_prod](https://leanprover-community.github.io/mathlib_docs/algebra/periodic.html#multiset.periodic_prod) it's used to derive the multiset result from the list result. I also attempted to use `finset.to_list` [once in my code](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Rational.20basis.20of.20a.20number.20field/near/271779420), but then I found the finset result already exists and there's no need to use the list result.

Both of these examples are in the "right direction" as you suggest. However, the current situation seems genuinely different, and we really need to derive the multiset result from the finset result, as explained by @xroblot in [the Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/.E2.9C.94.20fintype.20enumerating.20a.20multiset/near/286912084).

Reply:
Now that I think more deeply about it, another approach would be the following refactoring, namely to define `esymm_multiset` and use it to define `esymm`:
```
def multiset.esymm {R} [comm_semiring R] (s : multiset R) (n : ℕ) : R :=
((s.powerset_len n).map multiset.prod).sum

def mv_polynomial.esymm (σ) [fintype σ] (R) [comm_semiring R] (n : ℕ) : mv_polynomial σ R :=
(finset.univ.val.map X).esymm n
```
And we could then directly state and prove the multiset Vieta's formula; in fact, `multiset.prod_C_add_X_coeff` will simply be stated as:
```
lemma _root_.multiset.prod_C_add_X_coeff (s : multiset R) (k : ℕ) (h : k ≤ s.card) :
  (s.map (λ r, C r + X)).prod.coeff k = s.esymm (s.card - k) := sorry
```

`multiset.esymm` would have better functorial properties, e.g. [mv_polynomial.rename_esymm](https://leanprover-community.github.io/mathlib_docs/ring_theory/polynomial/symmetric.html#mv_polynomial.rename_esymm) requires renaming by an equivalence, but the corresponding `multiset.map_esymm` would work with any semiring homomorphism. This probably would allow us to golf some proofs. Is anyone interested in taking on such a refactoring?

Reply:
The refactoring seems like it would be nontrivial, for example there is no analogue of [finset.prod_add](https://leanprover-community.github.io/mathlib_docs/algebra/big_operators/ring.html#finset.prod_add); [multiset.prod_map_add](https://leanprover-community.github.io/mathlib_docs/data/multiset/antidiagonal.html#multiset.prod_map_add) is stated in terms of antidiagonal instead of powerset. The former was introduced in #2258 and used to prove the binomial theorem, and they didn't introduce a multiset version at the same time.

Reply:
Given the status quo, deriving the multiset result from the finset version is probably the quickest, so I think we should focus on that. What do people think about the [solution proposed above](https://github.com/leanprover-community/mathlib/pull/14908#discussion_r905398881)?

Reply:
@alreadydone My concern wasn't about computability -- having it be computable just a nice side-effect. Rather it's the fact that this function circumvents the induction principle of multisets -- using choice makes definitions less robust (what if across multiple definitions the same choices are made and you end up formalizing something other than you think you did?). It's ok if this function depends on `decidable_eq` since (1) if you care about computability you'll make sure you have such an instance and (2) if you don't you can use the classical decidability instances.

> the analogue of map_of_fin_card must be proved another way. Have you tried?

Yes. See `map_to_enum_finset` in https://gist.github.com/kmill/a72b75c781ace68a60fa9eb56de33cd2 for one approach, or `multiset.enumerate` right after it for an enumeration function.

(I don't have any comment about whether a multiset enumeration type is necessary for Vieta's formula.)

Reply:
@kmill You are certainly free to PR your enumeration type and function. If you do that, I'll switch to using your canonical `map_to_enum_finset` instead of my `map_of_fin_card` in the proof of `multiset.prod_C_add_X_coeff`. However, I'd like to keep my `(map_)of_fin_card` (renaming them to `(map_)enum_of_fin_card` and let them take arguments `{l : list α} (hl : (l : multiset α) = s)`, and change the docstring to say that it allows you to enumerate a multiset by `fin s.card` once you choose an ordering `l` on `s`. How does that sound?

Reply:
I pushed the change. The proof of `multiset.map_enum_of_fin_card` simplified a lot, due to the generalization from `s.to_list` to `l`, making `subst` possible. I think removing `decidable_eq` is an advantage, because for example `ℕ → ℕ` and `real` don't have decidable_eq, right? But it's conceivable that we might want to enumerate a multiset of functions or reals (e.g. roots of a real polynomial).

Reply:
Just a quick note to let you know that I have started working on the refactor mentioned above, that is define
```lean
def multiset.esymm {R} [comm_semiring R] (s : multiset R) (n : ℕ) : R :=
((s.powerset_len n).map multiset.prod).sum
```
and then prove directly (that is without using the `finset` version): 
```lean
lemma multiset.prod_C_add_X_coeff (s : multiset R) (k : ℕ) (h : k ≤ s.card) :
  (s.map (λ r, C r + X)).prod.coeff k = s.esymm (s.card - k) := sorry
```
I am still far for being done though but I am hoping I'll have a first version soon.

Reply:
@xroblot  @alreadydone so what is the situation here? This PR and #15008 use different approaches for Vieta's formulas, so we should decide which one to keep. Of course we can keep certain results in both PR, but which one should prove Vieta?

Reply:
If I understand correctly, if we go ahead with #15008, we should add a corollary in that PR saying
```lean
lemma prod_X_add_C_eq_sum_esymm (s : multiset R) :
  (s.map (λ r, C r + X)).prod.coeff (s.card - j) = s.esymm j :=
```
and then copy to that PR the lemmas `_root_.multiset.prod_X_sub_C_coeff`, `vieta` from this PR.

Reply:
@riccardobrasca This PR adds 70 lines (+107 -37), among which 37 lines are dedicated to proof the X_sub_C version and the polynomial.root version, so **33 lines** is used to prove the multiset version of Vieta (i.e. `multiset.prod_C_add_X_coeff`, which is definitionally what @hrmacbeth wants except `j` and `s.card - j` are switched). #15008 uses **139 lines** (+146 -7) to do the same (almost; `prod_X_add_C_eq_sum_esymm` there is an equality of polynomials, short of extracting the coefficients).

I'm not sure how much #15008 can be golfed down; the multiset proof added there seems to be in parallel to the finset proof, using `multiset.esymm` instead of `mv_polynomial.esymm` (so why should the lemmas be put in files under *polynomial*?), and the PR doesn't redefine `mv_polynomial.esymm` in terms of `multiset.esymm` and doesn't use the multiset version to prove the finset version, which likely led to a lot of code duplication. In contrast, this PR uses the finset version to prove the multiset version, which is why it can be so short.

The concern of @kmill seemed to be that the enumeration function was randomly chosen and noncanonical, but

+ I've solved the issue by taking a list as an argument and making it fully computable, and now I only invoke choice within a proof, in the final result `multiset.prod_C_add_X_coeff`. I could also make a computable `trunc` version without the list, but Kyle recommended against it [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/fin.20.28s.2Ecard.29.20finset.20equiv/near/288054990); I could also make it an existential statement (it's not a just a nonempty statement because there is a crucial property the enumeration function needs to be satisfied to connect between finset and multiset), but Kyle is [also not in favor of that](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/fin.20.28s.2Ecard.29.20finset.20equiv/near/288080065).

+ The recent PR #15080, which adds three randomly chosen enumeration function, was quickly approved by Kyle; If the same standard is applied to this PR, the enumeration function shouldn't be a problem.

+ I also agreed to use Kyle's enumeration function, once it's PR'd and merged into mathlib.

@eric-wieser 's concern was that going from finset to multiset is the "wrong direction", but I'd argue that it's better than code duplication. If #15008 successfully refactored all the finset versions to be easily derived from the multiset versions, I would support it, but that may be a substantially larger effort.

Can people express their latest thoughts after seeing and comparing both PRs? Should we poll on Zulip?

Reply:
> This PR adds 70 lines (+107 -37), among which 37 lines are dedicated to proof the X_sub_C version and the polynomial.root version, so **33 lines** is used to prove the multiset version of Vieta (i.e. `multiset.prod_C_add_X_coeff`, which is definitionally what @hrmacbeth wants except `j` and `s.card - j` are switched). #15008 uses **139 lines** (+146 -7) to do the same (almost; `prod_X_add_C_eq_sum_esymm` there is an equality of polynomials, short of extracting the coefficients).

@alreadydone I wasn't trying to imply that you lacked `multiset.prod_C_add_X_coeff`.  Rather, I was sketching the path to transfer the last portion of your PR into #15008, if that PR is selected:  from `prod_X_add_C_eq_sum_esymm` in that PR, the swapped version of `multiset.prod_C_add_X_coeff` is deduced, and then your last two lemmas can be copied over.

I don't know enough about multisets to have an opinion between the two PRs (but I agree that if #15008 is selected then the Vieta file needs some golfing).

Reply:
@alreadydone I'm sorry this process is seeming arbitrary and capricious. It's not fun for anyone when reviews get dragged out.

Regarding the opinions you're attributing to me, they are not accurate and the Zulip posts don't support them. A `trunc`-valued enumeration function would be fine. My concern is not computability: *we should avoid circumventing the quotient API in definitions*. The quotient API requires that definitions that use multisets provably are independent of the list representative. If you made a `trunc`-valued function, then that would preserve this property.  In proofs, you can circumvent the API all you want. (I'm not too happy that there is a `multiset.to_list` function -- at best we should be using `quotient.out` directly.)

Regarding PR 15080, many times it's easier to prove things about cardinalities by composing equivalences, and `fintype.equiv_fin` can be useful for that. The equivalences introduced in that PR are thin wrappers around them. Also, these are actual equivalences from finsets coerced to types -- multisets don't have any type coercion to support that yet.

(Regarding your earlier comment when you say "I think removing `decidable_eq` is an advantage", note that you can still use a function depending on `decidable_eq` even when equality isn't decidable because of the classical decidability instances. Choice is OK for them since they're singletons.)

I'll give you that `enum_of_fin_card` function is a clever solution, but it seems like it's working around API problems elsewhere. I haven't had time to dig into what's going on there though, unfortunately. However, it seems that you don't use `enum_of_fin_card` much except for `multiset.map_enum_of_fin_card`. Maybe a straightforward solution is to fold it into the lemma and use this lemma directly?
```lean
theorem multiset.map_univ_nth_le_to_list {α : Type*} (s : multiset α) :
  (finset.univ : finset (fin s.card)).val.map (λ i, s.to_list.nth_le i (s.length_to_list.symm ▸ i.2)) = s :=
begin
  have h := s.coe_to_list,
  generalize_hyp h' : s.to_list = l at h,
  subst h,
  convert congr_arg coe l.map_nth_le,
  rw ← multiset.coe_map,
  rcongr,
end
```

Reply:
@kmill I find the proof a bit too long and unnatural. If we simply inline the defintion `enum_of_fin_card` and keep the `l` and `h` arguments, we could keep the simple proof:
```
import data.fintype.basic
theorem multiset.map_univ_nth_le {α} (s : multiset α)
  {l : list α} (h : (l : multiset α) = s) :
  (finset.univ : finset (fin s.card)).val.map (λ i, l.nth_le i $ by { subst h, exact i.2 }) = s :=
by { subst h, exact congr_arg coe l.map_nth_le }
```
This would keep the final proof of `prod_C_add_X_coeff` unchanged except for the name (the invocation of `to_list` would remain in that proof).

We could also use an existential statement:
```
theorem multiset.exists_enum_of_fin_card {α} (s : multiset α) :
  ∃ f : fin s.card → α, finset.univ.val.map f = s :=
by { rcases s, exact ⟨λ i, s.nth_le i i.2, congr_arg coe s.map_nth_le⟩ }
```

The trunc version is also easy to make:
```
def multiset.trunc_enum_of_fin_card {α} (s : multiset α) :
  trunc {f : fin s.card → α // finset.univ.val.map f = s} :=
quotient.rec_on s (λ l, trunc.mk ⟨λ i, l.nth_le i i.2, congr_arg coe l.map_nth_le⟩)
  (λ _ _ _, subsingleton.elim _ _)
```

For the last two solutions, we'd just use `obtain` to get the enumeration function in proofs. I just pushed the trunc version because it's "stronger"; I'm not sure whether to include the exists version, since the trunc version is equally easy to use (just an additional pair of brackets). I see that you made two PRs to deliberately remove use of `fintype.trunc_equiv_fin` but that's when the arbitrarily chosen definition is already in mathlib, so I think my use of the trunc version in the proof is OK?

Reply:
@hrmacbeth Yes I understand you were just stating what you need (that are in this PR but missing in #15008). I was just explaining to Riccardo that it takes a lot more lines in #15008 to get to the results people care about (and that PR is "less controversial" probably because it just got out of draft status and no one has reviewed it).

@xroblot If my PR is merged, I think your `antidiagonal_powerset` is certainly worth keeping, but I'm not sure about the others. The [finset version](https://leanprover-community.github.io/mathlib_docs/data/finset/powerset.html#finset.powerset_len_eq_filter) of `multiset.finset.powerset_len_eq_filter` has a 1-line proof and that's probably because there's developed API for finset that doesn't exist for multiset, and the correct thing to do is to develop the API for multiset and derive the finset version from it, in my opinion. (Notice that `finset.powerset_len` is defined in using `multiset.powerset_len`, but `antidiagonal` hasn't got a finset version in mathlib.) If you follow the finset proof closely, then it should be possible to use your intermediate lemmas to golf the finset counterparts, but with your current approach, the lemmas `count_nil_powerset`, `bind_powerset_card`, and `bind_powerset_len` simply don't exist in the finset world, and may raise questions from reviewers whether they're truly necessary.

Reply:
I wrote #15008, following a suggestion in this thread, since I needed the result and I had the feeling the situation of this PR was a bit stuck. I agree that it doesn't make a lot of sense to do things this way without a complete refactor of `mv_polynomial.esymm` in terms of `multiset.esymm`, but I didn't want to do it right away since I was not sure #15008 would be accepted. Now, if everybody is happy with the latest changes to this PR and want to use it, that's fine with me. One reason I did #15008 was to see if I could do it and I learned quite a bit doing it. (I am still quite a newbie with `mathlib` which explains why my proofs are far from optimal).

Reply:
From a quick glance, I think #15008 looks like a good approach. I'll take a longer look in a few days.

Reply:
@alreadydone I got some time to develop out a multiset enumeration type (#15094) to help out here. It's fine if you don't end up using it (however, it would be a nice test to see whether it's up to the task).

Re your version of `multiset.map_univ_nth_le`, sure that's fine. I don't think there's anything wrong with my proof, but I'm certainly not going to force you to use it :-)

Re `trunc` vs an existential, both seem fine but may as well go with the `trunc` one. (And yes, the reason I switched those uses of `fintype.trunc_equiv_fin` over was that the noncomputable versions already exist and they simplified the proofs a bit -- but it might be good to review mathlib someday and see why we're using `fintype.equiv_fin`.)

Reply:
What is the situation about this? I really think we want Vieta's formulas, in one form on another.

Reply:
@xroblot Now that #15008 is merged, I guess I can close this? If anything here is useful, feel free to extract to another PR.

Reply:
I think there's still a bunch of useful low-level API lemmas in this PR @alreadydone. I haven't checked to see if any of the vieta stuff in this PR is worth keeping.

Reply:
OK, it seems that PR doesn't connect to `polynomial.roots` and do the negation part. I guess I'll adapt my code to use the new `multiset.esymm` API then.